src/streamlit_extras/prometheus/__init__.py

```diff
@@ -0,0 +1,136 @@
+        appropriately formatted OpenMetrics text encoding for all the stored metrics.
+
+        Then do a bit of string manipulation to package it in the format expected
+        by Streamlit's stats handler, so the final output looks the way we expect.
+        """
+        DUPLICATE_SUFFIX = "\n# EOF\n"
+        output_str = generate_latest(self.registry).decode(encoding="utf-8")
+        if not output_str.endswith(DUPLICATE_SUFFIX):
+            raise ValueError("Unexpected output from OpenMetrics text encoding")
+        output = CustomStat(metric_str=output_str[: -len(DUPLICATE_SUFFIX)])
+        return [output]
+
+
+@extra
+def streamlit_registry() -> CollectorRegistry:
+    """
+    Expose Prometheus metrics (https://prometheus.io) from your Streamlit app.
+
+    Create and use Prometheus metrics in your app with `registry=streamlit_registry()`.
+    The metrics will be exposed at Streamlit's existing `/_stcore/metrics` route.
+
+    See more example metrics in the Prometheus Python docs:
+    https://prometheus.github.io/client_python/
+
+    Note: To produce accurate metrics, you are responsible to ensure that unique metric
+    objects are shared across app runs and sessions. We recommend either 1) initialize
+    metrics in a separate file and import them in the main app script, or 2) initialize
+    metrics in a cached function (and ensure the cache is not cleared during execution).
+
+    For an app running locally you can view the output with
+    `curl localhost:8501/_stcore/metrics` or equivalent.
+    """
+    from streamlit.runtime import Runtime
+
+    stats = Runtime.instance().stats_mgr
+
+    # Did we already register it elsewhere? If so, return that copy
+    for prv in stats._cache_stats_providers:
+        if isinstance(prv, PrometheusMetricsProvider):
+            return prv.registry
+
+    # This is the function was called, so create the registry
+    # and hook it into Streamlit stats
+    registry = CollectorRegistry(auto_describe=True)
+    prv = PrometheusMetricsProvider(registry=registry)
+    stats.register_provider(prv)
+    return registry
+
+
+def example():
+    import streamlit as st
+    from prometheus_client import Counter
+
+    @st.cache_resource
+    def get_metric():
+        registry = streamlit_registry()
+        return Counter(
+            name="my_counter",
+            documentation="A cool counter",
+            labelnames=("app_name",),
+            registry=registry,  # important!!
+        )
+
+    SLIDER_COUNT = get_metric()
+
+    app_name = st.text_input("App name", "prometheus_app")
+    latest = st.slider("Latest value", 0, 20, 3)
+    if st.button("Submit"):
+        SLIDER_COUNT.labels(app_name).inc(latest)
+
+    st.write(
+        """
+        View a fuller example that uses the (safer) import metrics method at:
+        https://github.com/sfc-gh-jcarroll/streamlit-extras/tree/prometheus-example/prometheus_example
```

Comment:
Don't forget to update this to point to the main repo

Reply:
Yeah until now I guess no example has really needed multi files. I don't think it would cause any issue to have that `prometheus_example` directory live within the `streamlit-extras/prometheus/` directory! You could call it simply `example/`